Add `fish_job_summary`, called whenever a job ends, stops, or is signalled

soumya92 opened the pull request:
## Description

Add `fish_job_summary`, called whenever a job ends, stops, or is signalled

This allows users to customise the behaviour of the shell by redefining the function. This is similar to how fish_title or fish_greeting behave, where the default implementation can be easily overridden.

The function receives as arguments the job id, command line, signal name and signal description.

Fixes #4319, maybe #1018 and #2727.

## TODOs:
<!-- Just check off what what we know been done so far. We can help you with this stuff. -->
- [ ] Changes to fish usage are reflected in user documentation/manpages.
- [X] Tests have been added for regressions fixed
- [X] User-visible changes noted in CHANGELOG.md

ridiculousfish commented:
Do not worry about the CI failure, there is one flaky test.

ridiculousfish commented on src/proc.cpp:
```diff
@@ -453,40 +453,34 @@ static void process_mark_finished_children(parser_t &parser, bool block_ok) {
-        // No truncation necessary.
-        return cmd;
+/// Call the fish_job_summary function with the given args.
+static void print_job_summary(parser_t &parser, wcstring_list_t args) {
```
`args` should have type `const wcstring_list_t &args` to avoid a copy

soumya92 replied:
Is it that easy to tell I don't use C++ elsewhere? :)
Done.

ridiculousfish commented on src/proc.cpp:
```diff
@@ -453,40 +453,34 @@ static void process_mark_finished_children(parser_t &parser, bool block_ok) {
+    wcstring buffer = wcstring(L"fish_job_summary");
+    for (const wcstring &arg : args) {
+        buffer.push_back(L' ');
+        buffer.append(escape_string(arg, ESCAPE_ALL));
```
I think escaping is not desirable here - the function should receive the actual command that was run. It can be escaped with `string escape` if the function chooses to (maybe it should).

soumya92 replied:
This is building a single string for execution, so the individual arguments need to be escaped once in order to be split up correctly.

With escaping the execution string will be `fish_job_summary 3 sleep\ 10\ \&', which will correctly provide the args "3" and "sleep 10 &". Without escaping the execution string will be `fish_job_summary 3 sleep 10 &`, which in addition to providing the wrong args ""3", "sleep", "10", will also try to background the function call.

You can see from the expect based tests that the special characters are not escaped when the function prints them, so I think that covers what you want.

soumya92 commented:
> Do we want to invoke fish_job_summary for all jobs and let it filter out foreground ones?

I like that idea! It would work very well for me, since I can then avoid duplicated code between this function and postexec. I can also simplify my own postexec which currently has special handling for "^C" on the foreground job.

A flag indicating fg vs bg would also very much help with customisations, since bg means the function could need to handle an existing prompt on the screen. (It could also defer printing the line until next prompt to avoid interleaving with command output, or print some more obvious borders to make it clear that the output is coming from a background job rather than whatever is running).

I will add another commit to layer that functionality.

soumya92 commented:
Can you take a look at this PR?

I've fixed the tests (mostly, clang is still failing for some reason I don't fully understand), and I think I've addressed the comments raised earlier.

soumya92 commented on share/functions/fish_job_summary.fish:
```diff
@@ -0,0 +1,52 @@
+function fish_job_summary -a job_id -a is_foreground -a cmd_line -a signal_or_end_name -a signal_desc -a proc_pid -a proc_name
```
Learned something new today! Done. Now to go update all my local configuration...

soumya92 commented on src/proc.cpp:
```diff
@@ -453,40 +453,35 @@ static void process_mark_finished_children(parser_t &parser, bool block_ok) {
-    fflush(stdout);
-    outp.flush_to(STDOUT_FILENO);
+static void print_job_status(parser_t &parser, const job_t *j, job_status_t status) {
+    wcstring_list_t args;
```
:+1:  Done

soumya92 commented on share/functions/fish_job_summary.fish:
```diff
@@ -0,0 +1,52 @@
+
+    set max_cmd_len 32
+    if test (string length $cmd_line) -gt $max_cmd_len
+        set truncated_len (math $max_cmd_len - (string length $ellipsis))
```
Done.

soumya92 commented on share/functions/fish_job_summary.fish:
```diff
@@ -0,0 +1,52 @@
+        set ellipsis \u2026
+    end
+
+    set max_cmd_len 32
```
Done.

soumya92 commented on share/functions/fish_job_summary.fish:
```diff
@@ -0,0 +1,52 @@
+    end
+
+    set -l ellipsis '...'
+    if string match -iqr 'utf.?8' -- (locale charmap)
```
That's probably better even without the forking concern. I know fish reads LANG elsewhere, so it makes sense to be consistent.

soumya92 commented:
Done. Increasing sleep times seems to have fixed the tests as well, thanks!

krobelus commented:
Splendid, thank you!
